mathics/builtin/lists.py

```diff
@@ -4097,3 +4105,129 @@ def apply_manual_k(self, p, k, evaluation, options):
+        'DistanceFunction': 'Automatic',
+        'Method': '"Scan"',
+    }
+
+    messages = {
+        'amtd': '`1` failed to pick a suitable distance function for `2`.',
+        'list': 'Expected a list or a rule with equally sized lists at position 1 in ``.',
+        'nimp': 'Method `1` is not implemented yet.',
+    }
+
+    rules = {
+        'Nearest[list_, pattern_]': 'Nearest[list, pattern, 1]',
+        'Nearest[pattern_][list_]': 'Nearest[list, pattern]',
+    }
+
+    def apply(self, items, pivot, limit, evaluation, options):
+        'Nearest[items_List, pivot_, limit_, OptionsPattern[%(name)s]]'
+
+        method = self.get_option(options, 'Method', evaluation)
+        if not isinstance(method, String) or method.get_string_value() != 'Scan':
+            evaluation('Nearest', 'nimp', method)
+            return
+
+        dist_p, repr_p = _dist_repr(items)
+
+        if dist_p is None or len(dist_p) != len(repr_p):
+            evaluation.message(self.get_name(), 'list', Expression('Nearest', items, x, n))
+            return
+
+        if limit.has_form('List', 2):
+            up_to = limit.leaves[0]
+            py_r = limit.leaves[1].to_mpmath()
+        else:
+            up_to = limit
+            py_r = None
+
+        if isinstance(up_to, Integer):
+            py_n = up_to.get_int_value()
+        elif up_to.get_name() == 'System`All':
+            py_n = None
+        else:
+            return
+
+        if not dist_p or (py_n is not None and py_n < 1):
+            return Expression('List')
+
+        multiple_x = False
+
+        distance_function_string, distance_function = self.get_option_string(
+            options, 'DistanceFunction', evaluation)
+        if distance_function_string == 'Automatic':
+            from mathics.builtin.tensors import get_default_distance
+
+            distance_function = get_default_distance(dist_p)
+            if distance_function is None:
+                evaluation.message(self.get_name(), 'amtd', 'Nearest', Expression('List', *dist_p))
+                return
+
+            if pivot.get_head_name() == 'System`List':
+                _, depth_x = walk_levels(pivot)
+                _, depth_items = walk_levels(dist_p[0])
+
+                if depth_x > depth_items:
+                    multiple_x = True
+
+        def nearest(x):
+            calls = [Expression(distance_function, x, y) for y in dist_p]
+            distances = Expression('List', *calls).evaluate(evaluation)
+
+            if not distances.has_form('List', len(dist_p)):
+                raise ValueError()
+
+            py_distances = [(_to_real_distance(d), i) for i, d in enumerate(distances.leaves)]
+
+            def pick():
+                if py_n is None:
+                    candidates = sorted(py_distances)
+                else:
+                    candidates = heapq.nsmallest(py_n, py_distances)
+
+                for d, i in candidates:
+                    if py_r is not None and d > py_r:
```

Comment:
I think we should filter before finding the `py_n` smallest:

1. Avoids comparison with `None` in heapq.nsmallest
2. If k elements are within the limit distance the runtime is (very) slightly better: O(n) + O(min(k, m) log(k)) vs O(m log (n)) for m nearest elements.